radixdlt/src/main/java/com/radixdlt/consensus/SyncQueues.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * (C) Copyright 2020 Radix DLT Ltd
+ *
+ * Radix DLT Ltd licenses this file to you under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied.  See the License for the specific
+ * language governing permissions and limitations under the License.
+ */
+
+package com.radixdlt.consensus;
+
+import com.google.common.collect.ImmutableCollection;
+import com.google.common.collect.ImmutableMap;
+import com.radixdlt.counters.SystemCounters;
+import com.radixdlt.counters.SystemCounters.CounterType;
+import com.radixdlt.crypto.ECPublicKey;
+import com.radixdlt.crypto.Hash;
+import io.reactivex.rxjava3.annotations.Nullable;
+import java.util.LinkedList;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Synchronous queuing mechanism for consensus events which require syncing
+ * before being able to effectively process it.
+ *
+ * A separate queue is created for each node in order to keep the message ordering invariant.
+ *
+ * This class is NOT thread-safe.
+ */
+public final class SyncQueues {
+	private final ImmutableMap<ECPublicKey, SyncQueue> queues;
+
+	private final SystemCounters counters;
+
+	public SyncQueues(
+		Set<ECPublicKey> nodes,
+		SystemCounters counters
+	) {
+		this.queues = nodes.stream().collect(ImmutableMap.toImmutableMap(n -> n, n -> new SyncQueue()));
+		this.counters = Objects.requireNonNull(counters);
+	}
+
+	class SyncQueue {
+		private final LinkedList<RequiresSyncConsensusEvent> queue;
+
+		private SyncQueue() {
+			this.queue = new LinkedList<>();
+		}
+
+		/**
+		 * If a vertexId is supplied, checks the top of the queue to see if
+		 * the event corresponds to the vertexId. If so, returns it.
+		 *
+		 * TODO: cleanup interfaces
+		 *
+		 * @param vertexId the vertexId to check, if null, no vertexId is checked
+		 * @return the top of the queue if requirements are met
+		 */
+		@Nullable
+		public RequiresSyncConsensusEvent peek(@Nullable Hash vertexId) {
```

Comment:
Not a big fan of the peek/pop approach used here.

Both the peek and pop methods here could be folded into a nullable `RequiresSyncConsensusEvent poll(...)` method, and just avoid the multistage peek/pop approach used in clients.

Reply:
Problem with poll is that we may need to put the event back into the front of the queue if its execution fails. Peek then pop when needed seems cleaner to me over poll then put if needed. Plus its more overhead.